frugen.c

```diff
@@ -177,21 +177,47 @@ fru_field_t * fru_encode_custom_binary_field(const char *hexstr)
 static
 bool json_fill_fru_area_fields(json_object *jso, int count,
                                const char *fieldnames[],
-                               char *fields[])
+                               typed_field_t *fields[])
 {
 	int i;
 	json_object *jsfield;
 	bool data_in_this_area = false;
 	for (i = 0; i < count; i++) {
-		json_object_object_get_ex(jso, fieldnames[i], &jsfield);
-		if (jsfield) {
-			const char *s = json_object_get_string(jsfield);
-			debug(2, "Field %s '%s' loaded from JSON",
-			         fieldnames[i], s);
-			fru_loadfield(fields[i], s);
-			data_in_this_area = true;
-		}
-	}
+		if (json_object_object_get_ex(jso, fieldnames[i], &jsfield)) {
+			// field is an object
+			json_object *typefield, *valfield;
+			if (json_object_object_get_ex(jsfield, "type", &typefield) &&
+			    json_object_object_get_ex(jsfield, "val", &valfield)) {
+				// expected subfields are type and val
+				const char *type = json_object_get_string(typefield);
+				const char *val = json_object_get_string(valfield);
+				if (!strcmp("binary", type)) {
+					debug(1, "Binary format not yet implemented");
+				} else if (!strcmp("bcdplus", type)) {
+					fields[i]->type = FIELD_TYPE_BCDPLUS;
+				} else if (!strcmp("6bitascii", type)) {
+					fields[i]->type = FIELD_TYPE_SIXBITASCII;
+				} else if (!strcmp("text", type)) {
+					fields[i]->type = FIELD_TYPE_TEXT;
+				} else {
+					debug(1, "Unknown type %s for field '%s'",
+					      type, fieldnames[i]);
+					continue;
+				}
+				fru_loadfield(fields[i]->val, val);
+				debug(2, "Field %s '%s' (%s) loaded from JSON",
+				      fieldnames[i], val, type);
+				data_in_this_area = true;
+                        } else {
```

Comment:
It's great that you've preserved compatibility with the old file format! Thank you!